src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -0,0 +1,196 @@
+        Map.entry(RangeQueryBuilder.class, List.of(obj -> ((RangeQueryBuilder) obj).fieldName())),
+        Map.entry(RegexpQueryBuilder.class, List.of(obj -> ((RegexpQueryBuilder) obj).fieldName())),
+        Map.entry(
+            SpanNearQueryBuilder.SpanGapQueryBuilder.class,
+            List.of(obj -> ((SpanNearQueryBuilder.SpanGapQueryBuilder) obj).fieldName())
+        ),
+        Map.entry(SpanTermQueryBuilder.class, List.of(obj -> ((SpanTermQueryBuilder) obj).fieldName())),
+        Map.entry(TermQueryBuilder.class, List.of(obj -> ((TermQueryBuilder) obj).fieldName())),
+        Map.entry(TermsQueryBuilder.class, List.of(obj -> ((TermsQueryBuilder) obj).fieldName())),
+        Map.entry(WildcardQueryBuilder.class, List.of(obj -> ((WildcardQueryBuilder) obj).fieldName()))
+    );
+
+    public static String buildShape(SearchSourceBuilder source, Boolean showFields) {
+        StringBuilder shape = new StringBuilder();
+        shape.append(buildQueryShape(source.query(), showFields));
+        shape.append(buildAggregationShape(source.aggregations(), showFields));
+        shape.append(buildSortShape(source.sorts(), showFields));
+        return shape.toString();
+    }
+
+    static String buildQueryShape(QueryBuilder queryBuilder, Boolean showFields) {
+        if (queryBuilder == null) {
+            return "";
+        }
+        QueryShapeVisitor shapeVisitor = new QueryShapeVisitor();
+        queryBuilder.visit(shapeVisitor);
+        return shapeVisitor.prettyPrintTree("", showFields);
+    }
+
+    static String buildAggregationShape(AggregatorFactories.Builder aggregationsBuilder, Boolean showFields) {
+        if (aggregationsBuilder == null) {
+            return "";
+        }
+        StringBuilder aggregationShape = recursiveAggregationShapeBuilder(
+            aggregationsBuilder.getAggregatorFactories(),
+            aggregationsBuilder.getPipelineAggregatorFactories(),
+            new StringBuilder(),
+            0,
+            showFields
+        );
+        return aggregationShape.toString();
+    }
+
+    static StringBuilder recursiveAggregationShapeBuilder(
+        Collection<AggregationBuilder> aggregationBuilders,
+        Collection<PipelineAggregationBuilder> pipelineAggregations,
+        StringBuilder outputBuilder,
+        int indentCount,
+        Boolean showFields
+    ) {
+        String baseIndent = TWO_SPACE_INDENT.repeat(indentCount);
+
+        //// Normal Aggregations ////
+        if (aggregationBuilders.isEmpty() == false) {
+            outputBuilder.append(baseIndent).append("aggregation:").append("\n");
+        }
+        List<String> aggShapeStrings = new ArrayList<>();
+        for (AggregationBuilder agg : aggregationBuilders) {
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.append(baseIndent).append(TWO_SPACE_INDENT).append(agg.getType());
+            if (showFields) {
+                if (agg instanceof ValuesSourceAggregationBuilder) {
```

Comment:
Don't need a map for agg similar to queries right?

Reply:
Let me see if that can be done. That will make the code more extensible.

Reply:
Done